File: typescript/src/payment/api/paymentInternalApi.ts
// API interne del Payment Bounded Context per setup e test.

import express, { type Express, type Request, type Response } from 'express';
import { getErrorMessage } from '@/common/api/errorMessage.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { RequestPayment, requestPayment } from '../application/commands/requestPayment.js';
import { PaymentFinder } from '../application/query/paymentFinder.js';
import { PaymentRequesting } from '../application/usecases/paymentRequesting.js';
import { PaymentId } from '../domain/payment/paymentId.js';
import { parseCreatePaymentRequest } from './createPaymentRequest.js';
import { toPaymentDetailsResponse } from './paymentDetailsResponse.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentInternalApi {
  private readonly _paymentRequesting: PaymentRequesting;
  private readonly _paymentFinder: PaymentFinder;

  constructor(paymentRequesting: PaymentRequesting, paymentFinder: PaymentFinder) {
    requireDependency(paymentRequesting, "paymentRequesting");
    requireDependency(paymentFinder, "paymentFinder");
    this._paymentRequesting = paymentRequesting;
    this._paymentFinder = paymentFinder;
  }

  configure(app: Express): void {
    const router = express.Router();
    router.use(express.json());

    router.post('/internals/payments', (req, res) => this.create(req, res));
    router.get('/internals/payments', (req, res) => this.getByClientReference(req, res));

    app.use(router);
  }

  /**
   * @swagger
   * /internals/payments:
   *   post:
   *     summary: Crea un pagamento (uso interno e test)
   *     tags: [Payment Internal]
   *     requestBody:
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/CreatePaymentRequest'
   *     responses:
   *       201:
   *         description: Pagamento creato
   *         headers:
   *           Location:
   *             description: URI del pagamento creato
   *             schema:
   *               type: string
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/PaymentDetailsResponse'
   *       400:
   *         description: Richiesta non valida
   */
  private create(req: Request, res: Response): void {
    let request;
    try {
      request = parseCreatePaymentRequest(req.body);
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    let paymentId: PaymentId;
    try {
      paymentId = new PaymentId(Uuid.fromString(request.paymentId));
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'invalid paymentId'));
      return;
    }

    let command: RequestPayment;
    try {
      command = requestPayment(
        paymentId,
        new ClientReference(Uuid.fromString(request.clientReference)),
        new Money(Number(request.amount)),
        new Timestamp(new Date(request.requestedAt))
      );
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'invalid request'));
      return;
    }

    try {
      this._paymentRequesting.invoke(command);
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    const details = this._paymentFinder.findDetailsById(paymentId);
    if (details === null) {
      res.status(500).send('payment not found after creation');
      return;
    }

    res.status(201)
      .set('Location', `/payments/${paymentId.value.value}`)
      .json(toPaymentDetailsResponse(details));
  }

  /**
   * @swagger
   * /internals/payments:
   *   get:
   *     summary: Cerca un pagamento per client reference
   *     tags: [Payment Internal]
   *     parameters:
   *       - in: query
   *         name: clientReference
   *         required: true
   *         schema:
   *           type: string
   *     responses:
   *       200:
   *         description: Pagamento trovato
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/PaymentDetailsResponse'
   *       400:
   *         description: Parametro clientReference mancante o non valido
   *       404:
   *         description: Pagamento non trovato
   */
  private getByClientReference(req: Request, res: Response): void {
    const clientReference = req.query.clientReference;
    if (clientReference === undefined || clientReference === null || typeof clientReference !== 'string' || clientReference.trim().length === 0) {
      res.status(400).send('clientReference query parameter is required');
      return;
    }

    let payment;
    try {
      payment = this._paymentFinder.findDetailsByClientReference(new ClientReference(Uuid.fromString(clientReference)));
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    if (payment === null) {
      res.status(404).send();
      return;
    }

    res.status(200).json(toPaymentDetailsResponse(payment));
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/payment/api/PaymentInternalApi.java
package it.giannibombelli.wsc2026.payment.api;

import io.javalin.config.JavalinConfig;
import io.javalin.http.Context;
import io.javalin.openapi.*;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.module.WebApi;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.RequestPayment;
import it.giannibombelli.wsc2026.payment.application.query.PaymentDetails;
import it.giannibombelli.wsc2026.payment.application.query.PaymentFinder;
import it.giannibombelli.wsc2026.payment.application.usecases.PaymentRequesting;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;

import java.util.Optional;
import java.util.UUID;

import static io.javalin.apibuilder.ApiBuilder.get;
import static io.javalin.apibuilder.ApiBuilder.post;
import static java.util.Objects.requireNonNull;

/**
 * Endpoint interni del Payment BC, separati da {@link PaymentApi} perché non fanno parte
 * del contratto pubblico.
 */
public final class PaymentInternalApi implements WebApi {

    private final PaymentRequesting paymentRequesting;
    private final PaymentFinder paymentFinder;

    public PaymentInternalApi(PaymentRequesting paymentRequesting, PaymentFinder paymentFinder) {
        this.paymentRequesting = requireNonNull(paymentRequesting);
        this.paymentFinder = requireNonNull(paymentFinder);
    }

    @Override
    public void configure(JavalinConfig config) {
        config.routes.apiBuilder(() -> {
            post("/internals/payments", this::create);
            get("/internals/payments", this::getByClientReference);
        });
    }

    @OpenApi(
        path = "/internals/payments",
        methods = HttpMethod.POST,
        summary = "Create a payment (internal use only)",
        tags = {"Payment Internal"},
        requestBody = @OpenApiRequestBody(content = {@OpenApiContent(from = CreatePaymentRequest.class)}),
        responses = {
            @OpenApiResponse(status = "201", content = {@OpenApiContent(from = PaymentDetailsResponse.class)}),
            @OpenApiResponse(status = "400")
        }
    )
    private void create(Context ctx) {
        CreatePaymentRequest request = parseCreatePaymentRequest(ctx);
        if (request == null) {
            return;
        }

        if (request.paymentId() == null) {
            ctx.status(400).result("paymentId is required");
            return;
        }
        if (request.clientReference() == null || request.clientReference().isBlank()) {
            ctx.status(400).result("clientReference is required");
            return;
        }
        if (request.amount() == null) {
            ctx.status(400).result("amount is required");
            return;
        }
        if (request.requestedAt() == null) {
            ctx.status(400).result("requestedAt is required");
            return;
        }

        PaymentId paymentId;
        try {
            paymentId = new PaymentId(request.paymentId());
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        RequestPayment command;
        try {
            command = new RequestPayment(
                paymentId,
                new ClientReference(UUID.fromString(request.clientReference())),
                new Money(request.amount()),
                new Timestamp(request.requestedAt())
            );
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        try {
            paymentRequesting.invoke(command);
        } catch (IllegalArgumentException | IllegalStateException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        PaymentDetails details = paymentFinder.findDetailsById(paymentId).orElse(null);
        if (details == null) {
            ctx.status(500).result("payment not found after creation");
            return;
        }

        ctx.status(201)
            .header("Location", "/payments/" + paymentId.value())
            .json(PaymentDetailsResponse.from(details));
    }

    @OpenApi(
        path = "/internals/payments",
        methods = HttpMethod.GET,
        summary = "Get a payment by client reference (internal use only)",
        tags = {"Payment Internal"},
        queryParams = {@OpenApiParam(name = "clientReference", required = true)},
        responses = {
            @OpenApiResponse(status = "200", content = {@OpenApiContent(from = PaymentDetailsResponse.class)}),
            @OpenApiResponse(status = "400"),
            @OpenApiResponse(status = "404")
        }
    )
    private void getByClientReference(Context ctx) {
        String clientReference = ctx.queryParam("clientReference");
        if (clientReference == null || clientReference.isBlank()) {
            ctx.status(400).result("clientReference query parameter is required");
            return;
        }

        Optional<PaymentDetails> payment;
        try {
            payment = paymentFinder.findDetailsByClientReference(new ClientReference(UUID.fromString(clientReference)));
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        if (payment.isEmpty()) {
            ctx.status(404);
            return;
        }

        ctx.status(200).json(PaymentDetailsResponse.from(payment.get()));
    }

    private CreatePaymentRequest parseCreatePaymentRequest(Context ctx) {
        try {
            return ctx.bodyAsClass(CreatePaymentRequest.class);
        } catch (Exception ex) {
            ctx.status(400).result("request body is required");
            return null;
        }
    }
}
